docs: decision record about API versioning (#3595)
eclipse-edc/Connector: docs/developer/decision-records/2023-11-09-api-versioning/README.md

# Consistent Versioning of external APIs

## Decision

Going forward, all our external APIs will be versioned consistently across all endpoints and controllers. This means,
that the

> Management API (to name a prominent example) will have the version encoded same URL within the same major version.

Reviewer: ```suggestion
Management API (to name a prominent example) will encode the same version number consistently across all its existing service endpoint URLs, within a major version.
```

> Management API (to name a prominent example) will have the version encoded same URL within the same major version.

Reviewer 2: ```suggestion
Management API (to name a prominent example) will have a common major version encoded URL.
```

## Rationale

Currently, every module that contributes an API can have its own versioning model. For example, at the time of writing
some APIs are available under `/v2/...`, some are available under `/v3/...` and some without a version specifier.

This can be confusing to developers who want to consume the API, because they not only need different base paths for
every API, they also need to re-check every endpoint when a new version of EDC is released, because every API can change
individually.

> Therefor we will use consistent versioning across _all_ client-facing APIs. At the time of writing, this includes just

Reviewer 2: ```suggestion
Therefore, we will use consistent versioning across _all_ client-facing APIs. At the time of writing, this includes just
```

the Management API.

## Approach

### Versioning scheme

Following the general semantics of SemVer, we will distinguish between a breaking change (major version bump) and
non-breaking and patch changes (minor/patch version bump). The major version will be represented in the URL:

```
https://some.host.com/api/management/v4/...
```

This means, client code only needs to be updated on major version changes of an API.

> Only major versions are represented in the URL

> Minor or patch updates will **not** be represented in the URL, and we will **not** host multiple minor/patch versions
> of an API at any given time.

Consequently, when multiple major versions of an API are available, every versioned URL (e.g. `/v3/assets`) effectively
contains the _latest_ version, so for instance `/v3/assets` would effectively point to version `3.1.4` of the Management
API.

> The same exact version is used for the entire Management API, we do **not** version independent modules or controllers

> The version of an API is independent of the version of EDC.

### Delegation mechanism for controllers

> Not every endpoint changes from version to version, that means we will have to make the _same_ controller available

Reviewer 2: ``
```suggestion
Not every endpoint changes from version to version. We will have to make the _same_ controller available
```

under _multiple_ URL paths as we do not want to duplicate controllers. There are several potential ways we could make
this work:

- Declare the `@Path` as [wildcard](https://docs.oracle.com/javaee/7/api/javax/ws/rs/Path.html)
- implement a custom interceptor, that parses the path and delegates to the correct controller

These are as yet untested and need further investigation with regard to the following aspects:

> - do the come up as individual endpoints in the OpenAPI documentation?

Reviewer 2: ```suggestion
- do controllers appear as individual endpoints in the OpenAPI documentation?
```

> - is there a significant impact w.r.t. resource consumption?

Reviewer: ```suggestion
- is there a significant impact w.r.t. resource consumption and response time?
```
